File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l, r = 0, len(height) - 1

        max_water = 0

        while l < r:
            containerWidth = r - l
            min_height = min(height[l], height[r])
            water = min_height * containerWidth
            max_water = max(max_water, water)

            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return max_water

File: 0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int maxWater = 0;
        
        while (l < r) {
            int containerWidth = r - l;
            int min_height = Math.min(height[l], height[r]);
            int water = min_height * containerWidth;
            
            maxWater = Math.max(maxWater, water);
            
            if (height[l] < height[r]) {
                l += 1;
            } else {
                r -= 1;
            }
        }
        return maxWater;
    }
}
